network_simulator/host.py

```diff
@@ -23,4 +27,65 @@ def __init__(self, address=None, flows={}, flow_packets_received={},
         self.flow_packets_received = flow_packets_received
         self.link = link
 
-    # TODO(team): More event subclasses to come?
\ No newline at end of file
+
+class HostReceivedPacketEvent(Event):
+    """
+    Host receives packets from a flow.
+    """
+
+    def __init__(self, host, packet):
+        """
+        Sanity check: make sure destination was this Host and that this isn't
+        a RouterPacket.
+        :ivar host: host that receives packet
+        :ivar packet: packet received
+        """
+        Event.__init__(self)
+
+        # Sanity check
+        assert packet.dest_id == host and \
+        packet.packet_id != ROUTER_PACKET_DEFAULT_ID
+
+        self.host = host
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Put packet in flow_packets_received and sort.
+        Update statistics.
+        :param main_event_loop: event loop where new Events will be scheduled.
+        :param statistics: statistics to update
+        """
+        # add packet to host.flow_packets_received
+        self.flow_packets_received[self.packet.flow_id] = self.packet.packet_id
```

Comment:
flow_packets_received is a map from flow ID to a sorted *list* of the packet IDs that have been received for that Flow. You cannot have this map from flow ID to an integer, which is what you are doing right now. If you don't understand this, please ask me.

So what you will want to do is something like:

```
old_packets_received = self.flow_packets_received.get(self.packet.flow_id, default=list())
old_packets_received.append(self.packet.packet_id)

# Only sort old_packets_received if it has more than one element and 
# the append added something out of order.
if len(old_packets_received) > 1 and old_packets_received[-1] < old_packets_received[-2]:
    old_packets_received = sorted(old_packets_received)

self.flow_packets_received[self.packet.flow_id] = old_packets_received
```

Here, old_packets_received is initially the old list of Packet IDs received for the given Flow ID. The "default" parameter in the get() call is used as a back-up in case this flow doesn't already have a corresponding element in flow_packets_received. I then add a Packet ID to this list. However, I want to make sure the list is sorted. This is done via the sorted() call. Then, finally, I update self.flow_packets_received's mapping (for the given flow ID) to the new list.

There is an optimization here in the if statement: we only need to sort if old_packets_received (after the append) has > 1 element and if the last element (old_packets_received[-1]) is less than the second-to-last element (old_packets_received[-2]), as this indicates the list is out of order in its last two elements (it must be in order for the others). If this does not make sense, ask me and I will explain it.